### DISCUSSION TITLE
VS Code Autocomplete

### DISCUSSION THREAD

**Discussion by godismyjudge95** [pestphp/pest#419] [Q&A] [answered]
Wondering if anyone has figured out a way to get VS Code autocomplete working with PHP Intelephense (paid version)?

I've spent about an hour googling and searching through various issues on github for pest and intelephense.  The only thing I could find that was somewhat relevant was this:  #5 

Based off of that issue I was able to get autocomplete working by adding this to the beginning of all my tests:
```
/** @var TestCase $this */
```

This seems like a really brute force method to get autocomplete working.  So I was wondering if the mentioned fix was ever merged:  https://github.com/pestphp/pest/issues/5#issuecomment-638465818

**Comment by nunomaduro** [comment, marked as answer]
@godismyjudge95 I think to only way to get autocomplete properly working on VS Code, is having someone developing a plugin.

**Reply by eduwass** [reply]
Here's the plugin: [eduwass/pest-autocomplete-support](https://github.com/eduwass/pest-autocomplete-support) 🤣

**Reply by johnRivs** [reply]
> Here's the plugin: [eduwass/pest-autocomplete-support](https://github.com/eduwass/pest-autocomplete-support) 🤣

Does it provide autocomplete for `$this`?

**Reply by ace-of-aces** [reply]
@johnRivs No, I just tried it, and it doesn't provide any better support for `$this` (you're still left with  the same error messages).

**Reply by ace-of-aces** [reply]
For anyone facing this issue: I've created a new package **IntelliPest** that solves exactly this.

It generates a helper file which patches support for `$this`, but also more:
- autocompletion for custom expectations
- supports multiple TestCases (that you may have configured through `Pest.php`)
- custom traits

Check it out: [ace-of-aces/intellipest](https://github.com/ace-of-aces/intellipest)

**Comment by johnRivs** [comment]
@godismyjudge95 How exactly did you get that working? Which `TestCase` are you importing and where are you putting it?

**Reply by godismyjudge95** [reply]
@johnRivs
There is a "base" test case in the tests folder that is initially there when Laravel is installed.  That is what the `TestCase` is referring to.
To get VS Code to autocomplete properly, I was putting the comment I mentioned above inside each test like so:

```
it('can autocomplete in vs code', function () {
    /** @var TestCase $this */
    
    $this->assertTrue(true);
});
```

By adding the comment to each of my tests I got proper autocompletion only inside the test itself.  Not sure if it is required to work but I also use [laravel ide helper](https://github.com/barryvdh/laravel-ide-helper).

**Reply by bn-l** [reply]
The full path to the namespace seems to work better:

```php
test('confirm password screen can be rendered', function () {
    /** @var Tests\TestCase $this */
    
    $user = User::factory()->create();

    $response = $this->actingAs($user)->get('/confirm-password');

    

    $response->assertStatus(200);
});
```

**Reply by ace-of-aces** [reply]
Hey, so I just created something that's basically like [Laravel IDE Helper](https://github.com/barryvdh/laravel-ide-helper), but for Pest.

It lets you get rid of all of these  `/** @var TestCase $this */` comments (and more): [ace-of-aces/intellipest](https://github.com/ace-of-aces/intellipest)

**Comment by eduwass** [comment]
Stumbled into this whilst developing a solution for this in v4, you might wanna check this out!
https://github.com/eduwass/pest-autocomplete-support

**Comment by luka-papez** [comment]
For anyone reading this 2 years later, I fixed the problem of `$this->...` autocomplete by adding more definitions to an "IDE helper file" (which we already had in our project) as mentioned here:

* https://gist.github.com/ace-of-aces/28502ff74ec8c875e00ea30451950150
* https://github.com/laravel/vs-code-extension/issues/524

**Reply by ace-of-aces** [reply]
Hey, glad you came across my work!

As of right now, I'm not sure if my PR to the Laravel VS Code Extension will be merged or not..

In the meantime, I've created a little composer package (a helper file generator) which works independently of Laravel and even VS Code: https://github.com/ace-of-aces/intellipest

Hope this helps!
